File: functions/create_stimuli.m
% create stimuli given parameters

function [stimuli, radius, aspect_ratio, angle] = create_stimuli(s_r, ar_r,or_r, imageSizeX, imageSizeY)

%      s: size
%     s_r: size range
%     ar: aspect ratio
%     ar_r: aspect ratio range
%     or: orientation
%     or_r: orientation range

    [columnsInImage rowsInImage] = meshgrid(1:imageSizeX, 1:imageSizeY);
    centerX = (imageSizeX+1)/2;
    centerY = (imageSizeY+1)/2;

    radius = randi( s_r); % radius
    aspect_ratio = ar_r(1) + (ar_r(2)-ar_r(1))*rand(1); % aspect ratio
    b = radius;
    a = radius*aspect_ratio;
    theta = randi( or_r)*pi/180; % orientation
    angle = round(180-theta*180/pi);

    % approach1
    img = double(( (columnsInImage - centerX)*cos(theta)+(rowsInImage - centerY)*sin(theta) ).^2/a^2 + ...
          ( (columnsInImage - centerX)*sin(theta)-(rowsInImage - centerY)*cos(theta) ).^2/b^2 ...
             <= 1)*150/255;
    img(img==0)=.5;

    sigma=.08;
    stimuli = img + sigma*randn(size(img));


    % % approach2
    % img_n2 = imnoise(img, 'gaussian', 0, sigma^2);
    % imshow(img_n2);
    % img_n2 = imnoise(img/255,'gaussian',0,(sigma/255)^2);
    % imshow(uint8(img_n2*255));

    % % approach3
    % img_n3 = awgn(img,var(img(:))/sigma);
    % imshow(uint8(img_n3));

end

File: functions/create_stimuli.py
import numpy as np

def get_stimuli(x1, x2, x3, imageSizeX, imageSizeY, sigma=.08):
    '''
    :param x1: feature1: length of the disk in pixels
    :param x2: feature2: width of the disk in pixels
    :param x3: feature2: angle of the orientation in degrees (45, 40, etc)
    :param imageSizeX: image x
    :param imageSizeY: image y
    :type sigma: sigma of external white noise
    :return: stimuli with white noise
    '''

    x = np.linspace(0, imageSizeX-1, imageSizeX).astype(int)
    y = np.linspace(0, imageSizeX-1, imageSizeY).astype(int)
    columnsInImage, rowsInImage = np.meshgrid(x, y)
    centerX = int((imageSizeX+1)/2)
    centerY = int((imageSizeY+1)/2)

    b = x1
    a = x2
    theta = (90-x3)*np.pi/180

    img = ( ( (columnsInImage - centerX)*np.cos(theta)+(rowsInImage - centerY)*np.sin(theta) )**2/a**2 + \
            ( ( columnsInImage - centerX)*np.sin(theta)-(rowsInImage - centerY)*np.cos(theta) )**2/b**2 <= 1).astype(float)\
          *140/255
    img[img==0]=.5
    stimuli = img + sigma*np.random.randn(img.shape[0],img.shape[1])
    return stimuli